### PR TITLE
Implemented the exclude column and order by feature in the DiffResult

### PR DESCRIPTION
**Pull request by StevenYWei**
Implemented the exclude column and order by feature in the DiffResult
Add a test for the new function in DiffResult.

### CODE REVIEW

**Comment by aosama** [comment]
I looked at the files change and I'm comfortable with the changes, perhaps @mmlinford and @matthewgillett can take a quick look too.

File: src/test/scala/org/finra/msd/sparkcompare/JdbcToJdbcTest.scala [hunk 1]
```diff
@@ -89,4 +102,11 @@ class JdbcToJdbcTest() extends SparkFunSuite {
     if (pair.inRightNotInLeft.count != 0)
       fail("Expected 0 differences coming from right table." + "  Instead, found " + pair.inRightNotInLeft.count + ".")
   }
+
+  test("testCompareAndSaveFile") {
+    deleteSavedFiles("testCompareAndSaveFile")
+    val noDiff = returnDiffWithSavingResult("Test1", "Test3", "testCompareAndSaveFile")
+    Assert.assertFalse("Expected differences. Instead found no difference!", noDiff)
```
Comment (matthewgillett) [inline src/test/scala/org/finra/msd/sparkcompare/JdbcToJdbcTest.scala]:
```suggestion
    Assert.assertFalse("Expected differences. Instead found no difference!", noDiff)
    
    val leftDiff = readSavedFile("testCompareAndSaveFile/inLeftNotInRight").split("\n")
    val rightDiff = readSavedFile("testCompareAndSaveFile/inRightNotInLeft").split("\n")
    if (leftDiff.length != 3) // Includes header line
      fail("Expected 2 differences coming from left table." + "  Instead, found " + leftDiff.length + ".")
    if (rightDiff.length != 3) // Includes header line
      fail("Expected 2 differences coming from right table." + "  Instead, found " + rightDiff.length + ".")

    val leftCols = leftDiff(0).split(",")
    val rightCols = rightDiff(0).split(",")
    if (leftCols.length != 4)
      fail("Expected 3 columns returned in left table differences." + "  Instead, found " + leftCols.length + ".")
    if (rightCols.length != 4)
      fail("Expected 3 columns returned in right table differences." + "  Instead, found " + rightCols.length + ".")

    if (leftDiff(0).contains("RIPENESS"))
      fail("Expected ripeness not to be included in left table differences.")
    if (rightDiff(0).contains("Ripeness"))
      fail("Expected ripeness not to be included in right table differences.")

    if (leftDiff(1).split(",")(1).toInt < leftDiff(2).split(",")(1).toInt)
      fail("Expected results to be sorted descending by price in left table differences.")
    if (rightDiff(1).split(",")(1).toInt < rightDiff(2).split(",")(1).toInt)
      fail("Expected results to be sorted descending by price in right table differences.")
```

File: src/test/scala/org/finra/msd/basetestclasses/SparkFunSuite.scala [hunk 2]
```diff
@@ -97,4 +100,9 @@ class SparkFunSuite
     }
   }
 
+  def deleteSavedFiles(folder : String): Unit = {
+    val path: Path = Path(outputDirectory + "/" + folder + "/")
+    Try(path.deleteRecursively())
+  }
+
```
Comment (matthewgillett) [inline src/test/scala/org/finra/msd/basetestclasses/SparkFunSuite.scala]:
```suggestion

def readSavedFile(folder : String): String = {
    val file = new java.io.File(outputDirectory + "/" + folder + "/")
      .listFiles
      .filter(_.isFile)
      .filter(_.getName.endsWith(".csv"))(0)

    scala.reflect.io.File(file).slurp()
  }
```

File: src/test/scala/org/finra/msd/basetestclasses/SparkFunSuite.scala [hunk 3]
```diff
@@ -27,6 +27,9 @@ import org.finra.msd.memorydb.MemoryDbHsql
 import org.finra.msd.sparkfactory.SparkFactory
 import org.scalatest._
 
+import scala.reflect.io.Path
```
Comment (matthewgillett) [inline src/test/scala/org/finra/msd/basetestclasses/SparkFunSuite.scala]:
```suggestion
+import scala.reflect.io.{File, Path}
```

File: src/test/scala/org/finra/msd/basetestclasses/SparkFunSuite.scala (hunk 2 above)
Comment (matthewgillett) [inline src/test/scala/org/finra/msd/basetestclasses/SparkFunSuite.scala]: Also update lines 76 and 77 with the following:
protected final def getTestResourceFile(file: String): java.io.File = {
  new java.io.File(getClass.getClassLoader.getResource(file).getFile)

File: src/test/scala/org/finra/msd/sparkcompare/JdbcToJdbcTest.scala (hunk 1 above)
Comment (StevenYWei) [reply src/test/scala/org/finra/msd/sparkcompare/JdbcToJdbcTest.scala]: Can you explain the the fail message in line 114, 116, 121 and 123?

Comment (matthewgillett) [reply src/test/scala/org/finra/msd/sparkcompare/JdbcToJdbcTest.scala]: Lines 114 and 116 are checking the number of rows returned in the diff file are as expected (there is a header line in the file, hence the checking of length 3 but message of expecting 2).
Lines 121 and 123 are checking the number of columns returned in the diff file are as expected (there is a repeated row count column in each row, hence the checking of length 4 but message of expecting 3).

Comment (StevenYWei) [reply src/test/scala/org/finra/msd/sparkcompare/JdbcToJdbcTest.scala]: I'm trying to make the fail message more clear.
```suggestion
    Assert.assertFalse("Expected differences. Instead found no difference!", noDiff)
    
    val leftDiff = readSavedFile("testCompareAndSaveFile/inLeftNotInRight").split("\n")
    val rightDiff = readSavedFile("testCompareAndSaveFile/inRightNotInLeft").split("\n")
    if (leftDiff.length != 3) // Includes header line
      fail("Expected 3 rows (1 header and 2 data) coming from left table." + "  Instead, found " + leftDiff.length + ".")
    if (rightDiff.length != 3) // Includes header line
      fail("Expected 3 rows (1 header and 2 data) coming from right table." + "  Instead, found " + rightDiff.length + ".")
      
    val leftCols = leftDiff(0).split(",")
    val rightCols = rightDiff(0).split(",")
    if (leftCols.length != 4)
      fail("Expected 4 columns (3 data columns and 1 repeated row count) returned in left table differences." + "  Instead, found " + leftCols.length + ".")
    if (rightCols.length != 4)
      fail("Expected 4 columns (3 data columns and 1 repeated row count) returned in right table differences." + "  Instead, found " + rightCols.length + ".")
      
    if (leftDiff(0).contains("RIPENESS"))
      fail("Expected ripeness not to be included in left table differences.")
    if (rightDiff(0).contains("Ripeness"))
      fail("Expected ripeness not to be included in right table differences.")
      
    if (leftDiff(1).split(",")(1).toInt < leftDiff(2).split(",")(1).toInt)
      fail("Expected results to be sorted descending by price in left table differences.")
    if (rightDiff(1).split(",")(1).toInt < rightDiff(2).split(",")(1).toInt)
      fail("Expected results to be sorted descending by price in right table differences.")
```

Comment (matthewgillett) [reply src/test/scala/org/finra/msd/sparkcompare/JdbcToJdbcTest.scala]: This looks good. Thanks for the suggestion and code update.

**Pull request merged**